File: Y2/POO/lab11/facade_ex.cpp
#include <iostream>
#include <string>

using namespace std;

class Monster{

    public:
    Monster(){
    }
    void spawn(){
        cout<<"Monster spawns in swamp.\n";
    }
};

class Environment{
    
    public:
    void generateMap(){
        cout<<"Map is generating.\n";
    }
};

class Player{
    int attackDmg=10;

    public:
    void spawn(){
        cout<<"Player spawns in the base.\n";
    }
};

class Game{
    Monster** monsters;
    Environment* environment;
    Player* player;

    public:
    Game (){
        monsters = new Monster*[10];
        environment = new Environment();
        player = new Player();
    }
    void startGame(){
        environment->generateMap();
        for(size_t i = 0; i < 10; i++)
            monsters[i]->spawn();
        player->spawn();
    }

    ~Game(){
        delete player;
        for(size_t i = 0; i < 10; i++)
            delete monsters[i];
        delete[] monsters;
        delete environment;
    }
};

int main(){
    Game game ;
    game.startGame();
}

File: Y2/POO/lab11/kotlin/facade_ex.kt
class Monster {
    fun spawn() {
        println("Monster spawns in swamp.")
    }
}

class Environment {
    fun generateMap() {
        println("Map is generating.")
    }
}

class Player {
    private val attackDmg = 10

    fun spawn() {
        println("Player spawns in the base.")
    }
}

class Game {
    private val monsters = Array(10) { Monster() }
    private val environment = Environment()
    private val player = Player()

    fun startGame() {
        environment.generateMap()
        for (monster in monsters) {
            monster.spawn()
        }
        player.spawn()
    }
}

fun main() {
    val game = Game()
    game.startGame()
}
